File: weird_captcha_gym/shared_runtime/server/incubator_graders/popup_exorcist.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "popup_exorcist"


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    challenge_id = str(ground_truth.get("challenge_id") or "")
    if str(payload.get("mechanic_id") or "") != MECHANIC_ID or str(ground_truth.get("mechanic_id") or "") != MECHANIC_ID:
        return {"graded": True, "passed": False, "feedback": "mechanic mismatch"}
    if not challenge_id or str(payload.get("challenge_id") or "") != challenge_id or str(public_state.get("challenge_id") or "") != challenge_id:
        return {"graded": True, "passed": False, "feedback": "stale challenge"}
    condition = ground_truth.get("control_condition")
    expected_sources: dict[str, str] | None = None
    if condition is not None:
        if public_state.get("control_condition") != condition:
            return {"graded": True, "passed": False, "feedback": "public control condition differs from containment contract"}
        if str(payload.get("task_id") or "") != str(ground_truth.get("task_id") or ""):
            return {"graded": True, "passed": False, "feedback": "controlled task identity mismatch"}
        interaction = str(condition.get("interaction") or "")
        expected_sources = {
            "simplified": {
                "focus": "window_select",
                "close": "selected_close_button",
                "resist": "selected_close_button",
                "drag": "selected_contain_button",
                "contain": "selected_contain_button",
            },
            "full": {
                "focus": "window_pointer",
                "close": "window_close_button",
                "resist": "window_close_button",
                "drag": "window_drag",
                "contain": "window_drag",
            },
        }.get(interaction)
        if expected_sources is None:
            return {"graded": True, "passed": False, "feedback": "containment interaction condition is invalid"}
    events = payload.get("events")
    if not isinstance(events, list) or not 5 <= len(events) <= 180:
        return {"graded": True, "passed": False, "feedback": "containment transcript is missing or outside limits"}
    originals = set(str(item) for item in ground_truth.get("popup_ids") or [])
    parasite = str(ground_truth.get("parasite_id") or "")
    echoes = set(str(item) for item in ground_truth.get("echo_ids") or [])
    parasite_ids = [str(item) for item in (ground_truth.get("parasite_ids") or [parasite])]
    infection_groups = {
        str(parent): [str(item) for item in group]
        for parent, group in (
            ground_truth.get("infection_groups")
            or {parasite: list(ground_truth.get("echo_ids") or [])}
        ).items()
    }
    containment_stages = list(
        ground_truth.get("containment_stages")
        or [ground_truth.get("containment") or {}]
    )
    stage_batches = [
        [str(item) for item in batch]
        for batch in (
            ground_truth.get("stage_batches")
            or [list(ground_truth.get("popup_ids") or [])]
        )
    ]
    maximum_resistance_strikes = int(ground_truth.get("maximum_resistance_strikes") or 3)
    stage_batch_sets = [set(batch) for batch in stage_batches]
    if (
        not parasite_ids
        or len(set(parasite_ids)) != len(parasite_ids)
        or not set(parasite_ids).issubset(originals)
        or set(infection_groups) != set(parasite_ids)
        or set(item for group in infection_groups.values() for item in group) != echoes
        or len(containment_stages) != len(parasite_ids)
        or len(stage_batches) != len(parasite_ids)
        or any(not batch for batch in stage_batch_sets)
        or set().union(*stage_batch_sets) != originals
        or sum(len(batch) for batch in stage_batch_sets) != len(originals)
        or any(
            batch.intersection(parasite_ids) != {parasite_ids[index]}
            for index, batch in enumerate(stage_batch_sets)
        )
    ):
        return {"graded": True, "passed": False, "feedback": "parasite strain contract is malformed"}
    parent_by_window = {
        window_id: parent
        for parent, group in infection_groups.items()
        for window_id in [parent, *group]
    }
    popup_by_id = {str(item.get("id")): dict(item) for item in public_state.get("popups") or []}
    positions = {
        window_id: [
            int(popup_by_id[window_id].get("x") or 0),
            int(popup_by_id[window_id].get("y") or 0),
        ]
        for window_id in originals
    }
    dimensions = {
        window_id: [
            int(popup_by_id[window_id].get("w") or 0),
            int(popup_by_id[window_id].get("h") or 0),
        ]
        for window_id in originals
    }
    live, infected = set(stage_batch_sets[0]), set()
    provoked: set[str] = set()
    contained_parents: set[str] = set()
    spawn_pending = ""
    stage_pending: int | None = None
    active_stage = 0
    contained = ""
    last_drag: dict[str, list[list[int]]] = {}
    resistance_strikes = 0
    purged = False
    for sequence, event in enumerate(events, start=1):
        if not isinstance(event, dict) or event.get("sequence") != sequence:
            return {"graded": True, "passed": False, "feedback": f"event {sequence} sequence mismatch"}
        kind, window_id = str(event.get("kind") or ""), str(event.get("window_id") or "")
        if spawn_pending and kind != "spawn":
            return {"graded": True, "passed": False, "feedback": "parasite echoes were not recorded immediately"}
        if stage_pending is not None and kind != "stage":
            return {"graded": True, "passed": False, "feedback": "next popup wave was not activated immediately"}
        if kind == "stage":
            stage_index = event.get("stage_index")
            activated_ids = [str(item) for item in event.get("activated_ids") or []]
            if (
                stage_pending is None
                or stage_index != stage_pending
                or event.get("input_source") != "containment_field"
                or activated_ids != stage_batches[stage_pending]
                or live.intersection(activated_ids)
            ):
                return {"graded": True, "passed": False, "feedback": "popup wave activation was forged"}
            live.update(activated_ids)
            active_stage = stage_pending
            stage_pending = None
            continue
        if kind == "focus":
            if expected_sources is not None and event.get("input_source") != expected_sources["focus"]:
                return {"graded": True, "passed": False, "feedback": "window focus used the wrong interaction input"}
            if window_id not in live:
                return {"graded": True, "passed": False, "feedback": "focused window was not live"}
            continue
        if kind == "close":
            if expected_sources is not None and event.get("input_source") != expected_sources["close"]:
                return {"graded": True, "passed": False, "feedback": "window close used the wrong interaction input"}
            if window_id not in live:
                return {"graded": True, "passed": False, "feedback": "closed window was not live"}
            if window_id in parasite_ids and window_id not in provoked:
                provoked.add(window_id)
                infected.add(window_id)
                spawn_pending = window_id
            elif window_id not in infected:
                live.remove(window_id)
            continue
        if kind == "spawn":
            listed = set(str(item) for item in event.get("echo_ids") or [])
            parent = str(event.get("parent_id") or "")
            expected_echoes = set(infection_groups.get(parent) or [])
            if not spawn_pending or parent != spawn_pending or listed != expected_echoes:
                return {"graded": True, "passed": False, "feedback": "parasite echo set was forged"}
            live.update(expected_echoes)
            infected.update(expected_echoes)
            parent_x, parent_y = positions[parent]
            parent_w, parent_h = dimensions[parent]
            for index, echo_id in enumerate(infection_groups[parent]):
                horizontal = (
                    74 + (index // 2) * 22
                    if index % 2
                    else -58 - (index // 2) * 18
                )
                positions[echo_id] = [
                    max(8, min(690 - parent_w, parent_x + horizontal)),
                    max(10, min(365 - parent_h, parent_y + 54 + index * 22)),
                ]
                dimensions[echo_id] = [parent_w, parent_h]
            spawn_pending = ""
            continue
        if kind == "drag":
            if expected_sources is not None and event.get("input_source") != expected_sources["drag"]:
                return {"graded": True, "passed": False, "feedback": "window drag used the wrong interaction input"}
            if window_id not in live:
                return {"graded": True, "passed": False, "feedback": "dragged window was not live"}
            samples = event.get("samples")
            if not isinstance(samples, list) or not samples or len(samples) > 80:
                return {"graded": True, "passed": False, "feedback": "window drag samples are missing"}
            clean = []
            for point in samples:
                if not isinstance(point, list) or len(point) != 2 or not all(isinstance(value, int) for value in point):
                    return {"graded": True, "passed": False, "feedback": "window drag sample is malformed"}
                if clean and math.hypot(point[0] - clean[-1][0], point[1] - clean[-1][1]) > 120:
                    return {"graded": True, "passed": False, "feedback": "window teleported during containment"}
                clean.append(point)
            if expected_sources is not None and expected_sources["drag"] == "window_drag":
                if len(clean) < 2:
                    return {
                        "graded": True,
                        "passed": False,
                        "feedback": "full window drag requires an anchored multi-sample path",
                    }
                expected_start = positions.get(window_id)
                if (
                    expected_start is None
                    or math.hypot(
                        clean[0][0] - expected_start[0],
                        clean[0][1] - expected_start[1],
                    )
                    > 3
                ):
                    return {
                        "graded": True,
                        "passed": False,
                        "feedback": "full window drag was not anchored to the visible window",
                    }
            positions[window_id] = list(clean[-1])
            last_drag[window_id] = clean
            continue
        if kind == "contain":
            if expected_sources is not None and event.get("input_source") != expected_sources["contain"]:
                return {"graded": True, "passed": False, "feedback": "window containment used the wrong interaction input"}
            parent = parent_by_window.get(window_id, "")
            if (
                not parent
                or parent not in provoked
                or parent in contained_parents
                or window_id not in infected
                or window_id not in live
                or window_id not in last_drag
            ):
                return {"graded": True, "passed": False, "feedback": "containment occurred before an infected window was physically discovered"}
            end = positions[window_id]
            width, height = dimensions[window_id]
            cx, cy = end[0] + width / 2, end[1] + height / 2
            well = containment_stages[len(contained_parents)]
            if not (float(well.get("x")) <= cx <= float(well.get("x")) + float(well.get("w")) and float(well.get("y")) <= cy <= float(well.get("y")) + float(well.get("h"))):
                return {"graded": True, "passed": False, "feedback": "infected window did not enter the visible containment well"}
            contained = window_id
            contained_parents.add(parent)
            if len(contained_parents) < len(parasite_ids):
                neutralized = {
                    *stage_batch_sets[active_stage],
                    *infection_groups[parent],
                }
                live.difference_update(neutralized)
                infected.difference_update(neutralized)
                stage_pending = len(contained_parents)
            continue
        if kind == "resist":
            if expected_sources is not None and event.get("input_source") != expected_sources["resist"]:
                return {"graded": True, "passed": False, "feedback": "parasite resistance used the wrong interaction input"}
            if not provoked or window_id not in infected or window_id not in live:
                return {"graded": True, "passed": False, "feedback": "resistance was recorded for a noninfected window"}
            resistance_strikes += 1
            if event.get("strike") is not None and event.get("strike") != resistance_strikes:
                return {"graded": True, "passed": False, "feedback": "resistance strike sequence mismatch"}
            if resistance_strikes >= maximum_resistance_strikes:
                return {"graded": True, "passed": False, "feedback": "maximum parasite resistance reached"}
            continue
        if kind == "purge":
            if (
                purged
                or not contained
                or len(contained_parents) != len(parasite_ids)
                or stage_pending is not None
                or active_stage != len(parasite_ids) - 1
                or str(event.get("contained_id") or "") != contained
            ):
                return {"graded": True, "passed": False, "feedback": "purge was not triggered by the contained parasite"}
            remaining = set(str(item) for item in event.get("remaining_before") or [])
            if remaining != live:
                return {"graded": True, "passed": False, "feedback": "purge inventory did not match the live desktop"}
            live.clear()
            purged = True
            continue
        return {"graded": True, "passed": False, "feedback": f"unknown containment event {kind}"}
    passed = (
        set(parasite_ids) == provoked == contained_parents
        and not spawn_pending
        and stage_pending is None
        and active_stage == len(parasite_ids) - 1
        and bool(contained)
        and purged
        and not live
    )
    return {
        "graded": True,
        "passed": passed,
        "feedback": (
            f"strains provoked={len(provoked)}/{len(parasite_ids)}; "
            f"strains contained={len(contained_parents)}/{len(parasite_ids)}; "
            f"live windows={len(live)}"
        ),
    }

File: weird_captcha_gym/shared_runtime/app/mechanics/popup_exorcist.js
(() => {
  "use strict";

  let model = null;
  const esc = (value) => String(value == null ? "" : value)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
  const COPY = {
    update: ["OPTIONAL UPDATE", "A component may be available."],
    coupon: ["BONUS WINDOW", "Terms continue elsewhere."],
    cleaner: ["DESKTOP CARE", "Several pixels require attention."],
    forecast: ["LOCAL FORECAST", "A strong chance of more windows."],
    player: ["MEDIA HELPER", "Playback has not been requested."],
    survey: ["ONE QUESTION", "Closing is also an answer."],
    prize: ["CLAIM PENDING", "No prize was actually described."],
  };

  function record(kind, details = {}) {
    const event = {sequence: model.events.length + 1, kind, ...details};
    model.events.push(event);
    return event;
  }

  function windowMarkup(item, infected = false) {
    const copy = COPY[item.theme] || COPY.update;
    const cue = item.anomaly_cue || "none";
    const stageIndex = Number(item.stage_index || 0);
    const stageClass = stageIndex > 0 ? "is-awaiting-stage" : "";
    const cueText = cue === "explicit" ? '<em class="anomaly-label">UNSTABLE PROCESS</em>' : "";
    return `<article class="parasite-window theme-${esc(item.theme)} ${infected ? "is-infected" : ""} ${stageClass}" data-window-id="${esc(item.id)}" data-behavior="${esc(item.runtime_behavior || "echo")}" data-strain-id="${esc(item.strain_id || "")}" data-stage-index="${stageIndex}" data-cue="${esc(cue)}" style="left:${item.x}px;top:${item.y}px;width:${item.w}px;height:${item.h}px;z-index:${item.z}">
      <header><i></i><span>${esc(item.title)}</span><button type="button" class="parasite-close" aria-label="Close">×</button></header>
      <div class="parasite-body"><b>${esc(copy[0])}</b><p>${esc(copy[1])}</p>${cueText}<div class="fake-progress"><i></i></div></div>
    </article>`;
  }

  function liveWindows() {
    return Array.from(document.querySelectorAll(
      ".parasite-window:not(.is-dead):not(.is-awaiting-stage)",
    ));
  }

  function updateCount() {
    const node = document.querySelector(".parasite-count b");
    if (node) node.textContent = String(liveWindows().length);
  }

  function selectedWindow() {
    if (!model.selectedId) return null;
    return document.querySelector(`.parasite-window[data-window-id="${CSS.escape(model.selectedId)}"]:not(.is-dead)`);
  }

  function updateControls() {
    document.querySelectorAll(".parasite-window").forEach((node) => {
      node.dataset.selected = String(node.dataset.windowId === model.selectedId && !node.classList.contains("is-dead"));
    });
    const selected = selectedWindow();
    const status = document.querySelector(".popup-selection-status");
    if (status) status.textContent = selected ? "WINDOW SELECTED" : "SELECT A VISIBLE WINDOW";
    const close = document.querySelector(".popup-close-selected");
    const contain = document.querySelector(".popup-contain-selected");
    if (close) close.disabled = !selected;
    if (contain) contain.disabled = !selected || model.provokedParents.size === 0;
  }

  function focus(node, inputSource) {
    if (
      !node
      || node.classList.contains("is-dead")
      || node.classList.contains("is-awaiting-stage")
    ) return;
    node.style.zIndex = String(++model.topZ);
    model.selectedId = node.dataset.windowId;
    record("focus", {window_id: node.dataset.windowId, input_source: inputSource});
    updateControls();
  }

  function updateFieldSignal() {
    const signal = document.querySelector(".field-signal");
    if (!signal) return;
    if (model.provokedParents.size === 0) {
      signal.textContent = "CONTAINMENT FIELD: DORMANT";
      return;
    }
    const progress = model.parents.length > 1
      ? ` · STRAINS ${model.containedParents.size}/${model.parents.length}`
      : "";
    signal.textContent = `CONTAINMENT FIELD: RESONANT${progress}`;
  }

  function updateResistance() {
    const readout = document.querySelector(".resistance-count");
    if (readout) readout.textContent = `${model.strikes}/${model.maximumResistanceStrikes}`;
  }

  function setContainmentStage(index) {
    const stage = model.containmentStages[Math.min(index, model.containmentStages.length - 1)];
    const well = document.querySelector(".containment-well");
    well.dataset.stageIndex = String(index);
    well.style.left = `${stage.x}px`;
    well.style.top = `${stage.y}px`;
    well.style.width = `${stage.w}px`;
    well.style.height = `${stage.h}px`;
  }

  function activateStage(index) {
    const ids = model.stageBatches[index] || [];
    ids.forEach((id) => {
      const node = document.querySelector(
        `.parasite-window[data-window-id="${CSS.escape(id)}"]`,
      );
      if (!node) return;
      node.classList.remove("is-awaiting-stage");
      node.classList.add("is-arriving");
      node.style.zIndex = String(++model.topZ);
      setTimeout(() => node.classList.remove("is-arriving"), 320);
    });
    record("stage", {
      stage_index: index,
      activated_ids: [...ids],
      input_source: "containment_field",
    });
    model.activeStage = index;
    const root = document.querySelector(".parasite-captcha");
    if (root) root.dataset.activeStage = String(index);
    setContainmentStage(index);
    updateCount();
    updateControls();
  }

  function activateContainment() {
    document.querySelector(".containment-well").dataset.active = "true";
    document.querySelector(".popup-field").classList.add("is-contaminated");
    updateFieldSignal();
    updateControls();
  }

  async function failDesktop() {
    if (model.submitting) return;
    model.submitting = true;
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify({
          mechanic_id: model.state.mechanic_id,
          task_id: model.state.task_id,
          challenge_id: model.state.challenge_id,
          events: model.events,
        }),
      });
      const outcome = await response.json();
      model.helpers.setReadout("FAIL", "error");
      setTimeout(() => outcome.state && model.helpers.render(outcome.state), 850);
    } catch (_error) {
      model.submitting = false;
      model.helpers.setReadout("FAIL", "error");
    }
  }

  function spawnEchoes(parent) {
    const field = document.querySelector(".popup-field");
    const parentId = parent.dataset.windowId;
    const echoIds = model.infectionGroups[parentId] || [];
    const base = {
      x: parseFloat(parent.style.left),
      y: parseFloat(parent.style.top),
      w: parent.offsetWidth,
      h: parent.offsetHeight,
      theme: parent.dataset.theme || "update",
      title: "DESKTOP MESSAGE",
    };
    echoIds.forEach((id, index) => {
      const horizontal = index % 2 ? 74 + Math.floor(index / 2) * 22 : -58 - Math.floor(index / 2) * 18;
      const item = {
        ...base,
        id,
        strain_id: parentId,
        x: Math.max(8, Math.min(690 - base.w, base.x + horizontal)),
        y: Math.max(10, Math.min(365 - base.h, base.y + 54 + index * 22)),
        z: ++model.topZ,
        runtime_behavior: "echo",
      };
      field.insertAdjacentHTML("beforeend", windowMarkup(item, true));
      const node = field.querySelector(`[data-window-id="${CSS.escape(id)}"]`);
      installWindow(node);
    });
    record("spawn", {
      parent_id: parentId,
      echo_ids: [...echoIds],
      input_source: "parasite_replication",
    });
    updateCount();
    updateControls();
  }

  async function submit(containedId) {
    if (model.submitting) return;
    model.submitting = true;
    record("purge", {
      contained_id: containedId,
      remaining_before: liveWindows().map((node) => node.dataset.windowId),
      input_source: "containment_field",
    });
    document.querySelector(".popup-field").classList.add("is-purging");
    liveWindows().forEach((node, index) => setTimeout(() => node.classList.add("is-dead"), index * 75));
    setTimeout(updateCount, 650);
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify({
          mechanic_id: model.state.mechanic_id,
          task_id: model.state.task_id,
          challenge_id: model.state.challenge_id,
          events: model.events,
        }),
      });
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.helpers.setReadout("PASS", "passed");
        document.querySelector(".parasite-captcha").classList.add("is-passed");
      } else {
        model.helpers.setReadout("FAIL", "error");
        setTimeout(() => outcome.state && model.helpers.render(outcome.state), 850);
      }
    } catch (_error) {
      model.submitting = false;
      model.helpers.setReadout("FAIL", "error");
    }
  }

  function tryContain(node, inputSource) {
    if (model.provokedParents.size === 0 || !node.classList.contains("is-infected")) return false;
    const parentId = node.dataset.strainId || node.dataset.windowId;
    const activeParent = model.parents[model.containedParents.size];
    if (
      parentId !== activeParent
      || !model.provokedParents.has(parentId)
      || model.containedParents.has(parentId)
    ) return false;
    const well = document.querySelector(".containment-well").getBoundingClientRect();
    const rect = node.getBoundingClientRect();
    const cx = rect.left + rect.width / 2;
    const cy = rect.top + rect.height / 2;
    if (cx < well.left || cx > well.right || cy < well.top || cy > well.bottom) return false;
    record("contain", {window_id: node.dataset.windowId, input_source: inputSource});
    node.classList.add("is-contained");
    model.containedParents.add(parentId);
    const complete = model.containedParents.size === model.parents.length;
    if (complete) {
      updateFieldSignal();
      submit(node.dataset.windowId);
      return true;
    }
    const completedBatch = new Set(
      model.stageBatches[model.containedParents.size - 1] || [],
    );
    document.querySelectorAll(".parasite-window").forEach((candidate) => {
      const strainId = candidate.dataset.strainId || candidate.dataset.windowId;
      if (strainId === parentId || completedBatch.has(candidate.dataset.windowId)) {
        candidate.classList.remove("is-contained");
        candidate.classList.add("is-dead");
      }
    });
    model.selectedId = null;
    activateStage(model.containedParents.size);
    model.helpers.setReadout(`STRAIN ${model.containedParents.size}/${model.parents.length} CONTAINED`, "active");
    updateCount();
    updateFieldSignal();
    updateControls();
    return true;
  }

  function performClose(node, inputSource) {
    if (!node || model.submitting || node.classList.contains("is-dead")) return;
    focus(node, inputSource === "window_close_button" ? "window_pointer" : "window_select");
    const behavior = node.dataset.behavior;
    const parentId = node.dataset.strainId || node.dataset.windowId;
    record("close", {window_id: node.dataset.windowId, input_source: inputSource});
    if (behavior === "replicate" && !model.provokedParents.has(parentId)) {
      if (parentId !== model.parents[model.containedParents.size]) return;
      model.provokedParents.add(parentId);
      node.classList.add("is-infected");
      activateContainment();
      spawnEchoes(node);
      return;
    }
    if (node.classList.contains("is-infected")) {
      node.classList.remove("pulse");
      void node.offsetWidth;
      node.classList.add("pulse");
      model.strikes += 1;
      record("resist", {
        window_id: node.dataset.windowId,
        strike: model.strikes,
        input_source: inputSource,
      });
      updateResistance();
      if (model.strikes >= Number(model.state.maximum_resistance_strikes || 3)) failDesktop();
      return;
    }
    node.classList.add("is-dead");
    if (model.selectedId === node.dataset.windowId) model.selectedId = null;
    updateCount();
    updateControls();
  }

  function performContainProxy(node) {
    if (!node || model.submitting || model.provokedParents.size === 0) return;
    focus(node, "window_select");
    const well = model.containmentStages[model.containedParents.size];
    const origin = [parseFloat(node.style.left), parseFloat(node.style.top)];
    const target = [
      Math.max(0, Math.min(700 - node.offsetWidth, well.x + well.w / 2 - node.offsetWidth / 2)),
      Math.max(0, Math.min(390 - node.offsetHeight, well.y + well.h / 2 - node.offsetHeight / 2)),
    ];
    const samples = Array.from({length: 13}, (_, index) => {
      const amount = index / 12;
      return [
        Math.round(origin[0] + (target[0] - origin[0]) * amount),
        Math.round(origin[1] + (target[1] - origin[1]) * amount),
      ];
    });
    node.style.left = `${target[0]}px`;
    node.style.top = `${target[1]}px`;
    record("drag", {
      window_id: node.dataset.windowId,
      samples,
      input_source: "selected_contain_button",
    });
    tryContain(node, "selected_contain_button");
  }

  function installWindow(node) {
    if (!node) return;
    node.dataset.theme = node.className.match(/theme-([^ ]+)/)?.[1] || "update";
    if (node.dataset.behavior === "replicate" && !node.dataset.strainId) {
      node.dataset.strainId = node.dataset.windowId;
    }
    node.dataset.selected = "false";
    node.addEventListener("pointerdown", (event) => {
      if (event.target.closest("button")) return;
      focus(node, model.interaction === "full" ? "window_pointer" : "window_select");
    });
    const header = node.querySelector("header");
    if (model.interaction === "full") {
      header.addEventListener("pointerdown", (event) => {
        if (event.target.closest("button")) return;
        event.preventDefault();
        header.setPointerCapture(event.pointerId);
        const start = [event.clientX, event.clientY];
        const origin = [parseFloat(node.style.left), parseFloat(node.style.top)];
        const samples = [[Math.round(origin[0]), Math.round(origin[1])]];
        const move = (moveEvent) => {
          const x = Math.max(0, Math.min(700 - node.offsetWidth, origin[0] + moveEvent.clientX - start[0]));
          const y = Math.max(0, Math.min(390 - node.offsetHeight, origin[1] + moveEvent.clientY - start[1]));
          node.style.left = `${x}px`;
          node.style.top = `${y}px`;
          const target = [Math.round(x), Math.round(y)];
          const previous = samples[samples.length - 1];
          const distance = Math.hypot(target[0] - previous[0], target[1] - previous[1]);
          const steps = Math.max(1, Math.ceil(distance / 48));
          for (let index = 1; index <= steps; index += 1) {
            samples.push([
              Math.round(previous[0] + (target[0] - previous[0]) * index / steps),
              Math.round(previous[1] + (target[1] - previous[1]) * index / steps),
            ]);
          }
        };
        const up = () => {
          header.removeEventListener("pointermove", move);
          header.removeEventListener("pointerup", up);
          header.removeEventListener("pointercancel", up);
          if (samples.length > 1) {
            record("drag", {
              window_id: node.dataset.windowId,
              samples: samples.slice(-80),
              input_source: "window_drag",
            });
            tryContain(node, "window_drag");
          }
        };
        header.addEventListener("pointermove", move);
        header.addEventListener("pointerup", up);
        header.addEventListener("pointercancel", up);
      });
    }
    const close = node.querySelector(".parasite-close");
    if (model.interaction === "full") {
      close.addEventListener("click", () => performClose(node, "window_close_button"));
    } else {
      close.disabled = true;
      close.tabIndex = -1;
    }
  }

  async function render(state, helpers) {
    document.body.dataset.mechanic = "popup-exorcist-v2";
    const interaction = state.control_condition?.interaction || "full";
    const parents = state.parasite_ids
      ? [...state.parasite_ids]
      : state.infection_groups
      ? Object.keys(state.infection_groups)
      : state.popups.filter((item) => item.runtime_behavior === "replicate").map((item) => item.id);
    const infectionGroups = state.infection_groups || {[parents[0]]: [...state.echo_ids]};
    const containmentStages = state.containment_stages || [state.containment];
    const stageBatches = state.stage_batches || [
      state.popups.map((item) => item.id),
    ];
    model = {
      state,
      helpers,
      interaction,
      events: [],
      topZ: 20,
      parents,
      infectionGroups,
      containmentStages,
      stageBatches,
      activeStage: 0,
      provokedParents: new Set(),
      containedParents: new Set(),
      maximumResistanceStrikes: Number(state.maximum_resistance_strikes || 3),
      strikes: 0,
      selectedId: null,
      submitting: false,
    };
    const proxyControls = interaction === "simplified" ? `<aside class="popup-proxy-controls">
      <span>SELECTED WINDOW CONTROL</span>
      <b class="popup-selection-status">SELECT A VISIBLE WINDOW</b>
      <button type="button" class="popup-close-selected" disabled>CLOSE SELECTED</button>
      <button type="button" class="popup-contain-selected" disabled>MOVE SELECTED TO CONTAINMENT</button>
    </aside>` : "";
    const resistance = state.control_condition?.difficulty === 2
      ? ""
      : `<span class="resistance-readout">RESISTANCE <b class="resistance-count">0/${model.maximumResistanceStrikes}</b></span>`;
    helpers.app.innerHTML = `<section class="parasite-captcha" data-interaction="${esc(interaction)}" data-multistrain="${parents.length > 1}" data-active-stage="0" data-challenge-id="${esc(state.challenge_id)}">
      <header class="parasite-head"><div><span>DESKTOP CONTAINMENT / LIVE FIELD</span><h1>${esc(state.prompt)}</h1></div><div class="parasite-count">WINDOWS <b>${stageBatches[0].length}</b></div></header>
      <main class="parasite-main">
        <section class="popup-field"><div class="field-wallpaper"><i></i><span>LOCAL DESKTOP</span></div><div class="containment-well" data-active="false" data-stage-index="0" style="left:${state.containment.x}px;top:${state.containment.y}px;width:${state.containment.w}px;height:${state.containment.h}px"><i></i><b>⌁</b></div>${state.popups.map((item) => windowMarkup(item)).join("")}</section>
        ${proxyControls}
      </main>
      <footer class="parasite-foot"><div class="readout" data-status="idle"></div><div class="parasite-foot-status">${resistance}<span class="field-signal">CONTAINMENT FIELD: DORMANT</span></div></footer>
    </section>`;
    document.querySelectorAll(".parasite-window").forEach(installWindow);
    if (interaction === "simplified") {
      document.querySelector(".popup-close-selected").addEventListener("click", () => performClose(selectedWindow(), "selected_close_button"));
      document.querySelector(".popup-contain-selected").addEventListener("click", () => performContainProxy(selectedWindow()));
    }
    updateFieldSignal();
    updateResistance();
    updateControls();
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.popup_exorcist = {render, rootSelector: ".parasite-captcha"};
})();
